Low Power MAC Protocol (X-MAC)

jlwatson opened the pull request:
Hi everyone,

### Pull Request Overview

This pull request adds a low power MAC protocol as a capsule wrapper for drivers implementing the `kernel::hil::radio::Radio` interface. Specifically, this is an implementation of **X-MAC**, heavily based on the original work for the protocol that you can read all about [here](http://www.cs.cmu.edu/~andersoe/papers/xmac-sensys.pdf). In addition, support for putting the RF233 radio to sleep has been added.

The overall goal is to trade latency for decreased energy usage. X-MAC nodes put their radios to sleep for the vast majority of the time and wake intermittently to detect incoming data. Transmitters wake and begin sending a stream of small, strobed 'preamble' packets (currently implemented using reserved frame types). When the receiving node wakes and both receives and ACKs a preamble frame, the transmitter can immediately send the desired data frame and go back to sleep.

A couple of notes: as a result of the protocol, the radio is at full power only for a small time slice (given the current sleep/wake period values, 3.8% of the time), but transmission is slower. If the sleep period were set to, say, 60 seconds, it could take up to a minute to successfully transfer a data packet. Additionally, if a destination node is unresponsive to preambles, a `ReturnCode::ENOACK` is generated, which may need to be handled by the application. Finally, since nodes using X-MAC expect other nodes in the network to exhibit the same behavior (maximum sleep period, preamble ACKs, etc.), it cannot be used to communicate with non-XMAC-wrapped radios.

### Testing Strategy

Ran the `ip_sense` and `tests/ieee802154/radio_rx` clients on two Imix boards and checked for good packet reception.

Since X-MAC doesn't work with non-XMAC-enabled nodes (as noted above), and likely no one currently wants to endure higher latency while testing the radio for their own uses, I haven't submitted changes to the `imix` board configuration. If someone wants to try this out, I've provided the necessary configuration [here](https://gist.github.com/jlwatson/a3ceff5f7c7dffe41cca71e70c5de30d) and described the necessary changes in `capsules/src/xmac.rs`. The timer delay calculations require floating point ops and a matching change to `Xargo.toml` (included in the gist).

### TODO

- [ ] More rigorous testing of the 'no-preamble-random-backoff' transmission optimization. That's still pending because I'll have to grab a third `imix` to test with simultaneous traffic from multiple nodes.
- [ ] There's currently an expectation in the `XMAC` receive handler that if a packet is received immediately before `radio::Radio::stop()` is called (i.e the radio has a pending sleep directly following the completion of the reception), it will cancel the pending sleep in favor of remaining awake for incoming (data) packets. Obviously, that's *slightly* not what one might expect from trying to stop the radio, so I'll explore getting around this somehow.
- [ ] Changing the current, static sleep period to a dynamic value adjusted to traffic load. This is a major part of the [original X-MAC analysis](http://www.cs.cmu.edu/~andersoe/papers/xmac-sensys.pdf) that would make the protocol more optimally energy efficient.

### Documentation Updated

- [X] Kernel: The relevant files in `/docs` have been updated or no updates are required.
- [X] Userland: The application README has been added, updated, or no updates are required.

### Formatting

- [X] `make formatall` has been run.

bradjc commented:
Cool to see this getting added to Tock! Hopefully this isn't a problem, but I split part of this PR into #694. That should make it easier to look over the changes to the existing RF233 code and we can merge that without it potentially blocking on other changes.

bradjc commented:
Rebased+squashed.

If i'm not mistaken, the relevant stack looks like:

```
           ieee802154::mac::Mac
+-----------------------------------------+
|                                         |
|          ieee802154::mac::MMacDevice    |
|                                         |
+-----------------------------------------+
            hil::radio::Radio
+-----------------------------------------+
|                                         |
|                xmac                     |
|                                         |
+-----------------------------------------+
            hil::radio::Radio
+------------------------------------------+
|                                          |
|                 RF233                    |
|                                          |
+------------------------------------------+
```

Does it make sense that the low power mac layer provides a `hil::radio` interface? It seems like maybe there should be an intermediate interface that isn't just the radio interface.

jlwatson commented:
I think you may be right. For context, I’ve implemented the low power layer here as a `hil::radio::Radio` interface here for a couple reasons: 
* Any low power protocol is invisible to the `ieee802154` layers above it: they just see a `hil::radio::Radio`, albeit one that acts weirdly and experiences high latency. It avoids breaking existing higher-level code by implementing the expected backend interface.
* It’s also really easy to _not_ have a low-power layer since just deleting the XMAC-related lines in the board configuration and using the raw RF233 will work just fine.

At the same time, it’s not explicit where a low power layer is expected or allowed. You could imagine attempting an arbitrarily tall stack of "low power" protocols, which is certainly not the intended use case but not disallowed. More importantly, the XMAC layer is now tied to the `hil::radio::Radio` interface and the [802.15.4 TRD](https://github.com/helena-project/tock/blob/master/doc/reference/trd-radio.md), which it honestly already tends to break. For example, `hil::radio::Radio`s aren't _really_ supposed to lie when asked if the radio is currently on (as is the case with XMAC). It'd be nice to explicitly recognize that some sort of `LowPowerMac::is_on()` implementation can say the radio is on even when that is not the case. Finally, extending the interface in the future is going to be much easier if it doesn't involve modifying the HIL.

Given that, here’s what I think a potential low power interface should draw from, if anyone has any thoughts:
* The configuration interface should be similar or identical to the corresponding `ieee802154::mac::Mac` functions. That includes essentially all of the `hil::radio::RadioConfig` trait except for the power configuration (i.e `reset()`, `start()`, and `stop()`). The **low power** layer should retain that control exclusively. We don't want a situation where XMAC is twiddling the power state of the radio and suddenly the networking stack comes in and turns off the whole radio.
* The transmission interface would remain very close to the `hil::radio::RadioData` interface, albeit under a different name. The low power layer shouldn’t contain any logic for preparing and encoding the MAC frames in the first place (and in a simple pass-through implementation with no low power protocol, that’s pretty much necessary). What we’re interested in on this level is a raw MAC frame in a byte buffer that we can transmit using the radio.

alevy commented:
@jlwatson Why is it that XMAC wouldn't be interchangeable with (rather than below) the `MacDevice` (which should probably renamed to something more specific)?

jlwatson commented:
@alevy Right now, the `MacDevice` is responsible for encoding the correct MAC frame and then passing it to the radio for transmission. It's the _radio_ that implements the media access protocol; in the case of the RF233, it's CSMA/CA. I think that's a good separation of concerns to maintain.

The XMAC capsule assumes that it is handed valid, encoded MAC frames - that functionality is left completely to the `MacDevice` - and focuses on implementing a media access protocol that (1) ensures that the receiver is awake and (2) calls into the radio for collision avoidance. Replacing the `MacDevice` would require copying (and maintaining) all of the encoding logic into XMAC and any other potential low power protocols. Maybe the naming of the `Mac` stack components aren't quite right, but I think that we still want to maintain that separation.

bradjc commented:
> It’s also really easy to not have a low-power layer

I agree it's nice to easily be able to remove the low power layer by simply omitting it, but it also shouldn't be hard to write a simple `AlwaysOnMac` layer and use that instead.

In general, I think the cons of exposing the radio interface to the higher layers outweigh the benefits. Certainly the correctness issues where higher levels could turn on/off the radio (or think they are) are problematic, but it also makes it confusing to users and developers who can't use the interfaces to understand where things go in the stack.

I propose changing the stack to look like:

```
    net::154Device

         transmit(payload)
         set/get_address()
         set_pan()

+----------------------------+
|                            |
|     Ieee154Framer          |
|   (currently MacDevice)    |
|                            |
+----------------------------+

 net::mac

      transmit(full_154_packet)
      set/get_address()
      set_pan()

+----------------------------+
|                            |
|        xmac (or other mac) |
|                            |
+----------------------------+

       hil::radio::Radio

+----------------------------+
|                            |
|         rf233              |
|                            |
+----------------------------+
```

This makes controlling the channel, on/off state, and TX power the MAC layer's responsibility. Upper layers can still change the address, PAN ID, and of course send packets. The MAC layer expects fully formed packets that it can put on the radio. The Ieee802154Framer (what is now MacDevice) looks like it does now.

What exactly is in each interface might take time to iron out, but I don't think we want things that aren't radios to look like they are.

alevy commented:
@jlwatson is there anything currently blocking you on this?

jlwatson commented:
@alevy nothing blocking; just got held up over the break. I'm close to finishing the layer rework and then we can get this PR out of the way. Thanks for checking in!

jlwatson commented:
Clearly I've done something wrong with the rustup bump + rebase combo 🥇 . I'll fix in a bit.

jlwatson commented:
_quick question_: I'm much worse at attempting to squash my commits than I thought - what's the right sequence of commands to make that work? 

-----------------------------------------------------------------------------------------

Per Brad's suggestion, I've reworked the interfaces from the upwards-facing 802.15.4 interface down to the radio to avoid masquerading MAC protocols as `kernel::hil::radio::Radio`s when they clearly aren't while restricting the ability of the upper layers to control the power state of the radio. The changes I've implemented result in a stack that looks like this (open to suggestions/improvements):

```
             +-------------------------------------+ 
             |                                     |              
             |  ieee802154::driver::RadioDriver    |    
             |             (or other)              |      
             |                                     |     
             +------------------|------------------+   
                                |       
                                |        
 ieee802154::device::MacDevice  |   
                                |        
                                |             
             +------------------|------------------+-       
             |                                     |   
             |  ieee802154::virtual_mac::MacUser   |   
             |                                     |     
             +------------------|------------------+          
                                |               
                                |      
             +------------------|------------------+      
             |                                     |          
             |  ieee802154::virtual_mac::MuxMax    |  
             |                                     |        
             +------------------|------------------+   
                                |  configuration: address, PAN    
                                |  is_on  
 ieee802154::device::MacDevice  |  prepare_data_frame         
                                |  transmit prepared Frame by passing to encoder 
                                |       
             +------------------|------------------+      
             |                                     |      
             |      ieee802154::framer::Framer     |         
             |                                     |          
             +------------------|------------------+         
                                |                  
                                |  configuration: address, PAN    
          ieee802154::mac::Mac  |  is_on         
                                |  transmit fully encoded 802.15.4 frames  
                                |      
             +------------------|------------------+       
             |                                     |      
             |       ieee802154::xmac::XMac        |              
             |       (or other MAC protocol)       |                
             |                                     |           
             +------------------|------------------+         
                                |      
                                |  configuration: address, PAN, tx power, channel
     kernel::hil::radio::Radio  |  full power state management      
                                |  transmit fully encoded 802.15.4 frames   
                                |                
             +------------------|------------------+     
             |                                     |      
             |            rf233::RF233             |                   
             |                                     |           
             +-------------------------------------+      
```
The `ieee80154::mac::Mac` interface allows the higher layers of the stack to modify the radio address, PAN, and check on the MAC protocol/radio power state. The actual power management, tx power, and channel management is reserved for the MAC protocol (e.g. XMAC). Further, the interface expects fully generated 15.4 frames; generating the appropriate headers is the responsibility of the higher layers, in this case, a `ieee802154::framer::Framer`.

The `Framer`, and, higher up, the `MacUser` expose the `ieee802154::device::MacDevice` interface, which expose many of the same configuration options (address, etc.) but allow a client to explicitly encode header options into their buffers and send `Frame` objects that will be converted into raw, completely encoded `&'static mut [u8]`s before passing to the `Mac` layer.

alevy commented:
> quick question: I'm much worse at attempting to squash my commits than I thought - what's the right sequence of commands to make that work?

If you're trying to squash commits that branch from different subtrees in the log, it might be too tricky to be worth it. But in the "normal" case (i.e. I made a bunch of commits in series that I really want to be one commit):

```bash
$ git rebase -i [oldest commit i want to be part of the squashed commits]
```
That'll pop up an editor with the commit log. Replace all the commits lines you want to squash from `pick` to either `squash` or `fixup`.

alevy commented on boards/imix/src/main.rs:
```diff
@@ -89,6 +90,11 @@ static mut RF233_BUF: [u8; radio::MAX_BUF_SIZE] = [0x00; radio::MAX_BUF_SIZE];
+
+// The MAC layer also requires a buffer to receive packets into. Some MAC
+// implementations like XMac may require an additional buffer to buffer packets.
+static mut MAC_RX_BUF: [u8; radio::MAX_BUF_SIZE] = [0x00; radio::MAX_BUF_SIZE];
```
Why do we need a buffer _in addition_ to the radio rx buffer?

jlwatson replied:
You're right, that's not necessary. I've removed it.

alevy commented on capsules/src/ieee802154/xmac.rs:
```diff
@@ -0,0 +1,632 @@
+//! X-MAC protocol layer for low power 802.15.4 reception, intended primarily
+//! to manage an Atmel RF233 radio.
+//!
+//! TODO: Test no-preamble transmission with randomized backoff, requires 3
```
Can we move the `TODO`s to non-exposed comments (e.g. just `//`)? I think it reads strange in the docs. If there are specific limitations that a user should be aware of, maybe write those as prose, instead.

jlwatson replied:
Moved!

alevy commented on capsules/src/ieee802154/framer.rs:
```diff
@@ -0,0 +1,791 @@
+//! and automatic acknowledgement. Radio power management and channel selection
+//! is also passed down to the MAC control layer.
+//!
+//! TODO: Encryption/decryption
```
Ditto with moving `TODO`s to `//` comments

alevy commented:
Agree with @bradjc that at least @hudson-ayers should probably take a pass on this.

alevy commented:
@jlwatson side note that this turned out to be a very clean PR. Nice job!

hudson-ayers commented on capsules/src/ieee802154/xmac.rs:
```diff
@@ -0,0 +1,632 @@
+// Time the radio will continue to send preamble packets before aborting the
+// transmission and returning ENOACK. Should be at least as large as the maximum
+// sleep time for any node in the network.
+const PREAMBLE_TX_MS: u32 = 250;
```
Seems like this should probably be 251, though it would very rarely matter

jlwatson replied:
Nice catch - agreed that it will probably never matter, but I've updated it to 251 just in case.

alevy commented:
@jlwatson I think if you address @hudson-ayers and my minor comments I think we can merge

jlwatson commented:
Couple of minor changes:

- Several updates to commenting and documentation, including the `MacDevice` interface in the networking stack design doc.
- Removed the extra `MAC_RX_BUF` in `main.rs` for imix; it's not necessary.
- Updated the X-MAC `main.rs` and `Xargo.toml` configuration for testing, located [here](https://gist.github.com/jlwatson/a3ceff5f7c7dffe41cca71e70c5de30d).

Should be good to go!
